src/hotspot/share/opto/callGenerator.cpp

```diff
@@ -809,6 +813,424 @@ CallGenerator* CallGenerator::for_vector_reboxing_late_inline(ciMethod* method,
+                scoped_value_cache = addp1->in(AddPNode::Base)->uncast()->in(0)->as_Call();
+              } else {
+                assert(scoped_value_cache == addp1->in(AddPNode::Base)->uncast()->in(0), "");
+              }
+              assert(in->in(MemNode::Memory)->is_Proj() && in->in(MemNode::Memory)->in(0) == scoped_value_cache, "");
+              Node* addp2 = addp1->in(AddPNode::Address);
+              Node* offset1 = addp1->in(AddPNode::Offset);
+              intptr_t const_offset = offset1->find_intptr_t_con(-1);
+              BasicType bt = TypeAryPtr::OOPS->array_element_basic_type();
+              int shift = exact_log2(type2aelembytes(bt));
+              int header = arrayOopDesc::base_offset_in_bytes(bt);
+              assert(const_offset >= header, "");
+              const_offset -= header;
+
+              Node* index = kit.gvn().intcon(const_offset >> shift);
+              if (addp2->is_AddP()) {
+                assert(!addp2->in(AddPNode::Address)->is_AddP() &&
+                       addp2->in(AddPNode::Base) == addp1->in(AddPNode::Base),
+                       "");
+                Node* offset2 = addp2->in(AddPNode::Offset);
+                assert(offset2->Opcode() == Op_LShiftX && offset2->in(2)->find_int_con(-1) == shift, "");
+                offset2 = offset2->in(1);
+#ifdef _LP64
+                assert(offset2->Opcode() == Op_ConvI2L, "");
+                offset2 = offset2->in(1);
+                if (offset2->Opcode() == Op_CastII && offset2->in(0)->is_Proj() &&
+                    offset2->in(0)->in(0) == get_cache_iff) {
+                  ShouldNotReachHere();
+                  offset2 = offset2->in(1);
+                }
+#endif
+                index = kit.gvn().transform(new AddINode(offset2, index));
+              }
+
+              if (get_first_iff == nullptr) {
+                get_first_iff = c->as_If();
+                first_index = index;
+              } else {
+                assert(get_second_iff == nullptr, "");
+                get_second_iff = c->as_If();
+                second_index = index;
+              }
+            }
+          } else if (c->is_RangeCheck()) {
+            // Kill the range checks as they are known to always succeed
+            kit.gvn().hash_delete(c);
+            c->set_req(1, kit.gvn().intcon(1));
+            C->record_for_igvn(c);
+          } else if (c->is_CallStaticJava()) {
+            assert(slow_call == nullptr, "");
+            slow_call = c->as_CallStaticJava();
+            assert(slow_call->method()->intrinsic_id() == vmIntrinsics::_SVslowGet, "");
+          } else {
+            assert(c->is_Proj() || c->is_Catch(), "");
+          }
+          wq.push(c->in(0));
+        }
+      }
+      // get_first_iff/get_second_iff contain the first/second check we ran into during the graph traversal but they may
+      // not be the first/second one in execution order. Perform another traversal to figure out which is first.
```

Comment:
Why can this not be done in the first traversal, and why does this (down) traversal do the right thing?

Can we assert that `c` is always CFG?
Please mention in a comment that we are only traversing the same CFG nodes from the first traversal.

Reply:
> Why can this not be done in the first traversal, and why does this (down) traversal do the right thing?

The first traversal starts from the end of the method and follows control paths until it reaches the `Thread.scopedValueCache()` call. Given the shape of the method and that some paths may have been trimmed and end with an uncommon trap, it could reach either the first or the second if that probe the cache first.